src/e2e/java/teammates/e2e/cases/e2e/InstructorStudentListPageE2ETest.java

```diff
@@ -0,0 +1,131 @@
+                testData.students.get("Student3Course3"),
+                testData.students.get("Student4Course3"),
+        };
+
+        Map<String, StudentAttributes[]> courseHeaderToStudents = new HashMap<>();
+        courseHeaderToStudents.put(course1Header, studentsInCourse1);
+        courseHeaderToStudents.put(course2Header, studentsInCourse2);
+        courseHeaderToStudents.put(course3Header, studentsInCourse3);
+
+        // Expand all headers first
+
+        for (String header : courseHeaderToStudents.keySet()) {
+            listPage.clickCourseTabHeader(header);
+        }
+
+        listPage.verifyStudentDetails(courseHeaderToStudents);
+
+        ______TS("link: enroll page");
+
+        InstructorCourseEnrollPage enrollPage = listPage.clickEnrollStudents(course3Header);
+        enrollPage.verifyIsCorrectPage(course3.getId());
+
+        listPage = loginAdminToPage(listPageUrl, InstructorStudentListPage.class);
+        listPage.clickCourseTabHeader(course3Header);
+
+        ______TS("link: view student details page");
+
+        StudentAttributes studentToView = testData.students.get("Student1Course3");
+        String studentEmail = studentToView.getEmail();
+
+        InstructorCourseStudentDetailsViewPage studentDetailsViewPage =
+                listPage.clickViewStudent(course3Header, studentEmail);
+        studentDetailsViewPage.verifyIsCorrectPage(course3.getId(), studentEmail);
+        studentDetailsViewPage.closeCurrentWindowAndSwitchToParentWindow();
+
+        ______TS("link: edit student details page");
+
+        InstructorCourseStudentDetailsEditPage studentDetailsEditPage =
+                listPage.clickEditStudent(course3Header, studentEmail);
+        studentDetailsEditPage.verifyIsCorrectPage(course3.getId(), studentEmail);
+        studentDetailsEditPage.closeCurrentWindowAndSwitchToParentWindow();
+
+        ______TS("link: view all records page");
+
+        InstructorStudentRecordsPage studentRecordsPage =
+                listPage.clickViewAllRecords(course3Header, studentEmail);
+        studentRecordsPage.verifyIsCorrectPage(course3.getId(), studentToView.getName());
+        studentRecordsPage.closeCurrentWindowAndSwitchToParentWindow();
+
+        ______TS("action: delete student");
+
+        StudentAttributes studentToDelete = testData.students.get("Student3Course3");
+
+        listPage.deleteStudent(course3Header, studentToDelete.getEmail());
+
+        StudentAttributes[] studentsAfterDelete = {
+                testData.students.get("Student1Course3"),
+                testData.students.get("Student2Course3"),
+                testData.students.get("Student4Course3"),
+        };
+
+        listPage.verifyStudentDetails(course3Header, studentsAfterDelete);
+        verifyAbsentInDatastore(studentToDelete);
+
+    }
+
+    private String createHeaderText(CourseAttributes course) {
+        return String.format("[%s]: %s", course.getId(), course.getName());
```

Comment:
Same for this, I feel like the header formatting belongs more in the page object